ljv/src/main/java/org/atpfivt/ljv/GraphBuilder.java

```diff
@@ -232,4 +236,80 @@ public String generateDOT(Object obj) {
                 .append("}\n")
                 .toString();
     }
+
+    private void generateHeader() {
+        out.append("digraph Java {\n")
+                .append("\trankdir=\"")
+                .append(ljv.getDirection())
+                .append("\";\n")
+                .append("\tnode[shape=plaintext]\n");
+    }
+
+    String generateDOT() {
+        List<Object> roots = this.ljv.getRoots();
+        IdentityHashMap<Object, List<Object>> sameObjectMap = new IdentityHashMap<>();
+        List<AbstractMap.SimpleEntry<Object, Boolean>> pairs = roots.stream()
+                .map(root -> new AbstractMap.SimpleEntry<>(root, false))
+                .collect(Collectors.toList());
+        // create a map of object to its references
+       while(true) {
+           Optional<AbstractMap.SimpleEntry<Object, Boolean>> notProcessedPairOptional = findNotProcessedPair(pairs);
+           if (notProcessedPairOptional.isPresent()) {
+               AbstractMap.SimpleEntry<Object, Boolean> objectBooleanSimpleEntry = notProcessedPairOptional.get();
+               objectBooleanSimpleEntry.setValue(Boolean.TRUE);
+               List<AbstractMap.SimpleEntry<Object, Boolean>> references = findReferences(objectBooleanSimpleEntry, pairs);
+               List<Object> refList = references.stream()
+                       .map(AbstractMap.SimpleEntry::getKey)
+                       .collect(Collectors.toList());
+               // may be count is sufficient here check
+               sameObjectMap.put(objectBooleanSimpleEntry.getKey(), refList);
+           }
+           else{
+               break;
+           }
+       }
+       // create the diagram first
+       generateHeader();
+       for(var entry : sameObjectMap.entrySet()){
+            this.generateDotInternal(entry.getKey());
+       }
+       // link number of references
+        for (var entry : sameObjectMap.entrySet()) {
+            out.append("\t")
+                    .append(entry.getValue().size()+1) // including itself
+                    .append(" -> ")
+                    .append(dotName(entry.getKey()))
+                    .append("[label=\"")
+                    .append("\",fontsize=12];\n");
+        }
+        // may be clear roots?
+        this.ljv.getRoots().clear();
+        return out
+                .append("}\n")
+                .toString();
+
+    }
+
+    /**
+     *
+     * @param pair
+     * @param pairs
+     * @return all references of a object in th pair and set the values to processed
+     */
+    private List<AbstractMap.SimpleEntry<Object, Boolean>> findReferences(AbstractMap.SimpleEntry<Object, Boolean> pair, List<AbstractMap.SimpleEntry<Object, Boolean>> pairs) {
+        return pairs.stream()
+                .filter(p -> p.getValue() == Boolean.FALSE)
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [RC: Suspicious reference comparison of Boolean values (RC_REF_COMPARISON_BAD_PRACTICE_BOOLEAN)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#rc-suspicious-reference-comparison-of-boolean-values-rc-ref-comparison-bad-practice-boolean).
This method compares two Boolean values using the == or != operator. Normally, there are only two Boolean values (Boolean.TRUE and Boolean.FALSE), but it is possible to create other Boolean objects using the new Boolean(b) constructor. It is best to avoid such objects, but if they do exist, then checking Boolean objects for equality using == or != will give results than are different than you would get using .equals(...).

Reply:
Yes, you are right.  I overlooked it now it's corrected. let me know your thoughts

Reply:
Happy to help.